elements/pfe-datetime/test/pfe-datetime.spec.ts

```diff
@@ -0,0 +1,118 @@
+import { expect, html } from '@open-wc/testing';
+import { createFixture } from '@patternfly/pfe-tools/test/create-fixture.js';
+
+import { PfeDatetime } from '@patternfly/pfe-datetime';
+
+describe('<pfe-datetime>', function() {
+  it('should upgrade', async function() {
+    const element = await createFixture<PfeDatetime>(`<pfe-datetime></pfe-datetime>`);
+    expect(element, 'the <pfe-datetime> should be an instance of PfeDatetime')
+      .to.be.an.instanceof(customElements.get('pfe-datetime'))
+      .and
+      .to.be.an.instanceof(PfeDatetime);
+  });
+
+  it('should show a simple date format if just a datetime attribute is provided', async function() {
+    const element = await createFixture<PfeDatetime>(html`
+      <pfe-datetime id="simple" datetime="Mon Jan 2 15:04:05 EST 2006">
+        Mon Jan 2 15:04:05 EST 2006
+      </pfe-datetime>`);
+    const text = element.shadowRoot!.querySelector('span')!.textContent;
+
+    expect(text, 'should just show a simple date').to.equal('1/2/2006');
+  });
+
+  it('should show a formatted date', async function() {
+    const element = await createFixture<PfeDatetime>(html`
+      <pfe-datetime id="justdate"
+          type="local"
+          day="numeric"
+          month="long"
+          year="numeric"
+          locale="en-US"
+          datetime="Mon Jan 2 15:04:05 EST 2006">
+        Mon Jan 2 15:04:05 EST 2006
+      </pfe-datetime>`);
+    const text = element.shadowRoot!.querySelector('span')!.textContent;
+
+    expect(text, 'should show a formatted date').to.equal('January 2, 2006');
+  });
+
+  it('should show a formatted date with time', async function() {
+    const element = await createFixture<PfeDatetime>(html`
+      <pfe-datetime id="withtime"
+          type="local"
+          weekday="long"
+          month="short"
+          day="2-digit"
+          year="numeric"
+          hour="2-digit"
+          minute="2-digit"
+          second="2-digit"
+          locale="en-US"
+          datetime="Mon Jan 2 15:04:05 EST 2006">
+        Mon Jan 2 15:04:05 EST 2006
+      </pfe-datetime>`);
+    const { textContent } = element.shadowRoot!.querySelector('span')!;
+    const regex = /.*?, .*?, \d{4}, \d+:\d+:\d+ .{2}/;
+
+    expect(regex.test(textContent!), 'should show a formatted date with time')
+      .to.be.true;
+  });
+
+  it('should show a relative time since the date', async function() {
+    const element = await createFixture<PfeDatetime>(html`
+      <pfe-datetime id="yearsago" type="relative" datetime="Mon Jan 2 15:04:05 EST 2006">
+        Mon Jan 2 15:04:05 EST 2006
+      </pfe-datetime>`);
+    const text = element.shadowRoot!.querySelector('span')!.textContent;
```

Comment:
I haven't seen the `!` used like this before.  How does it differ from optional chaining with a `?`

Reply:
this is the [non-null assertion operator](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#non-null-assertion-operator-postfix-), a typescript exclusive. It's a postfix operator that modifies a union of anything with a nully type, removing the nully types, i.e. asserting that the expression is not nully.

in this case, `querySelector` returns `HTMLSpanElement|null`, so the `!` here asserts _a priori_ that the result of the expression is an `HTMLSpanElement`.

Unlike the optional chaining operator `?.` which performs a check at run time, the non-null assertion operator `!` has no runtime significance. In other words it's basically the same as `as HTMLSpanElement`, just a little terser. For that reason, our linting rules allow the non-null assertion operator **only** in test files.

Reply:
Unbelievable explanation.  Thank you 🙏